README.md
# DisqusService

Wrapper for Disqus APIs written in Swift.
This open-source library allows you to integrate it into your iOS apps.
Learn more about [Disqus APIs](https://disqus.com/api/docs/).

[![Version](https://img.shields.io/cocoapods/v/DisqusService.svg?style=flat)](http://cocoapods.org/pods/DisqusService)
[![Language Swift 3.0](https://img.shields.io/badge/Language-Swift%203.0-green.svg?style=flat)](https://swift.org)
[![Platforms iOS](https://img.shields.io/cocoapods/p/DisqusService.svg?style=flat)](http://www.apple.com/ios)
[![License MIT](https://img.shields.io/cocoapods/l/DisqusService.svg?style=flat)](https://github.com/ascarrambad/DisqusService/master/LICENSE)

## Installation

DisqusService is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'DisqusService'
```
## Usage

Add the following code in your `AppDelegate` class

```swift
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
    DisqusService.shared.set(publicKey: "<your-public-key>",
                             secretKey: "<your-secret-key>",
                             redirectURI: "<your-redirect-uri>")
}

@available(iOS 9.0, *)
func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool {
    if let sourceApplication = options[.sourceApplication] as? String {
        if sourceApplication == "com.apple.SafariViewService" {
            NotificationCenter.default.post(name: .DisqusServiceSafariAuthDidClose, object: url)
        }
    }
    return true
}
```

Because Disqus APIs don't support schemes to redirect your app after OAuth authentication, you MUST provide an URL which then redirects to the following scheme: `disqus-redirect://`.
To do so in Project Settings → Info: scroll down to **URL Types** and add an entry with `+`

- **Identifier**: your-app-bundle-id, i.e. `com.disqusService.testApp`
- **URL Schemes**: `disqus-redirect`

![](Assets/xcode.png)

Authenticating/Deauthenticating a user is very simple:
```swift
DisqusService.shared.authenticate(viewController: yourVC) { (success) in
    if success {
        
    } else {
        
    }
}

////////////////

DisqusService.shared.logout()
```

This library also includes a convenient class called DisqusComment to better handle received comments from a thread.
Here's an example:

```swift

let params = ["thread" : "<your-thread-id>", "order" : "asc"]
let posts: [DisqusComment] = []

DisqusService.shared.performGETRequest(api: "posts/list", authRequired: false, params: params) { (data, success) -> Void in
    if success {
        let posts = data?["response"] as! [[AnyHashable : Any]]

        for postJSON in posts {
            if let post = DisqusComment(disqusData: postJSON) {
                posts.append(post)
            }
        }
    }
}
```

Please note that certain APIs require authentication and MUST be specified.
API usage example:

```swift

let paramsDetails = ["forum" : "<your-forum>", "thread" : "<your-thread-id>"]
DisqusService.shared.performGETRequest(api: "threads/details", authRequired: false, params: paramsDetails) { (data, success) in

}

let paramsVote = ["post" : "<your-post-id>", "vote" : "1"]
DisqusService.shared.performPOSTRequest(api: "posts/vote", authRequired: true, params: paramsVote) { (data, success) in

}
```

## Author

Matteo Riva, matteoriva@me.com

## License

DisqusService is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Changelog

## Version 1.0.1

- Small bug fixes

## Version 1.0.0

- Initial release


Sources/DisqusUser.swift
//
//	DisqusUser.swift
//  DisqusService
//
//  Copyright (c) 2016 Matteo Riva <matteoriva@me.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.

import Foundation

internal class DisqusUser: NSObject, NSCoding {
    
    private(set) var accessToken: String
    private var refreshToken: String
    let userID: String
    let username: String
    var isTokenValid = true
    
    init?(json: [AnyHashable : Any]) {
        guard let accessToken = json["access_token"] as? String,
            let refreshToken = json["refresh_token"] as? String,
            let userID = json["user_id"] as? Int,
            let username = json["username"] as? String
            else { return nil }
        
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.userID = "\(userID)"
        self.username = username
    }
    
    required init?(coder aDecoder: NSCoder) {
        accessToken = aDecoder.decodeObject(forKey: "access_token") as! String
        refreshToken = aDecoder.decodeObject(forKey: "refresh_token") as! String
        userID = aDecoder.decodeObject(forKey: "user_id") as! String
        username = aDecoder.decodeObject(forKey: "username") as! String
    }
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(accessToken, forKey: "access_token")
        aCoder.encode(refreshToken, forKey: "refresh_token")
        aCoder.encode(userID, forKey: "user_id")
        aCoder.encode(username, forKey: "username")
    }
    
    func refreshToken(publicKey: String, secretKey: String, completionHandler: @escaping (Bool) -> Void) {
        
        let url = URL(string: "https://disqus.com/api/oauth/2.0/access_token/")!
        
        let params = ["grant_type" : "refresh_token",
                      "client_id" : publicKey,
                      "client_secret" : secretKey,
                      "refresh_token" : refreshToken]
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        var paramString = ""
        for (key,value) in params {
            paramString += "\(key)=\(value)&"
        }
        
        request.httpBody = paramString.data(using: String.Encoding.utf8)!
        
        URLSession.shared.dataTask(with: request, completionHandler: {[unowned self] (data, _, error) in
            if error == nil {
                if let json = (try? JSONSerialization.jsonObject(with: data!, options: [])) as? [AnyHashable : Any] {
                    
                    guard let accessToken = json["access_token"] as? String,
                        let refreshToken = json["refresh_token"] as? String
                        else {
                            DisqusService.shared.logout()
                            return
                    }
                    
                    self.accessToken = accessToken
                    self.refreshToken = refreshToken
                }
            }
            completionHandler(error == nil)
            self.isTokenValid = error == nil
            }).resume()
        
    }
}


Sources/DisqusComment.swift
//
//  DisqusComment.swift
//  DisqusService
//
//  Copyright (c) 2016 Matteo Riva <matteoriva@me.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.


import Foundation

public extension Notification.Name {
    static let DisqusServiceDidPostNewComment = Notification.Name("DisqusServiceDidPostNewComment")
}

public class DisqusComment: NSObject {
    
    public let profileImageURL: URL
    public let authorID: String?
    public let author: String?
    public let message: String
    public let postID: String
    public let isDeleted: Bool
    public let elapsedTime: String?
    public let points: Int
    
    public let parentID: String?
    public var replies = [DisqusComment]()
    
    required public init?(disqusData: [AnyHashable : Any]) {
        
        guard let author = disqusData["author"] as? [AnyHashable : Any],
            let avatar = author["avatar"] as? [AnyHashable : Any],
            let avatarURL = avatar["permalink"] as? String,
            let comment = disqusData["raw_message"] as? String,
            let tmpDate = disqusData["createdAt"] as? String,
            let isDeleted = disqusData["isDeleted"] as? Bool,
            let id = disqusData["id"] as? String,
            let points = disqusData["points"] as? Int
            else { return nil }
        
        self.profileImageURL = URL(string: avatarURL)!
        self.author = author["name"] as? String
        self.authorID = author["id"] as? String
        self.message =  comment
        self.isDeleted = isDeleted
        self.postID = id
        self.points = points
        
        if let parentID = disqusData["parent"] as? Int {
            self.parentID = "\(parentID)"
        } else {
            self.parentID = nil
        }
        
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy'-'MM'-'dd'T'HH':'mm':'ss Z"
        
        if let creationDate = formatter.date(from: "\(tmpDate) +0000") {
            let today = Date()
            let distanceBetweenDates = today.timeIntervalSince(creationDate)
            
            if (distanceBetweenDates/31536000 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/31536000)) y"
            } else if (distanceBetweenDates/2592000 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/2592000)) mes"
            } else if (distanceBetweenDates/604800 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/604800)) set"
            } else if (distanceBetweenDates/86400 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/86400)) gg"
            } else if (distanceBetweenDates/3600 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/3600)) h"
            } else if (distanceBetweenDates/60 >= 1) {
                elapsedTime = "\(Int(distanceBetweenDates/60)) min"
            } else if (distanceBetweenDates >= 30) {
                elapsedTime = "\(Int(distanceBetweenDates)) sec"
            } else {
                elapsedTime = "Ora"
            }
        } else {
            elapsedTime = "n/d"
        }
        
        super.init()
    }
    
    public func attachReplies(_ replies: [[AnyHashable : Any]]) {
        
        let real = replies.filter { $0["parent"] as! Int == Int(self.postID)! }
        let remaining = replies.filter { $0["parent"] as! Int != Int(self.postID)! }
        
        if real.count == 0 { return }
        
        for comment in real {
            if let dComment = DisqusComment(disqusData: comment) {
                self.replies.append(dComment)
                if remaining.count != 0 {
                    dComment.attachReplies(remaining)
                }
            }
        }
    }
   
}


Sources/DisqusAuthViewController.swift
//
//  DisqusAuthViewController.swift
//  DisqusService
//
//  Copyright (c) 2016 Matteo Riva <matteoriva@me.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.

import UIKit
import WebKit

internal class DisqusAuthViewController: UIViewController, WKNavigationDelegate {
    
    private let webView = WKWebView()
    
    var callback: ((String) -> Void)!
    var url: URL!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.addSubview(webView)
        webView.navigationDelegate = self
        webView.allowsBackForwardNavigationGestures = false
        webView.translatesAutoresizingMaskIntoConstraints = false
        let height = NSLayoutConstraint(item: webView, attribute: .height, relatedBy: .equal, toItem: view, attribute: .height, multiplier: 1, constant: 0)
        let width = NSLayoutConstraint(item: webView, attribute: .width, relatedBy: .equal, toItem: view, attribute: .width, multiplier: 1, constant: 0)
        view.addConstraints([height, width])
        
        let endButton = UIBarButtonItem(barButtonSystemItem: .done,
                                        target: self,
                                        action: #selector(dismissAction))
        endButton.tintColor = UIColor.black
        
        navigationController?.navigationBar.tintColor = UIColor.black
        navigationItem.leftBarButtonItem = endButton
        navigationItem.title = "Disqus Login"
        
        webView.load(URLRequest(url: url!))
    }

    func dismissAction() {
        dismiss(animated: true, completion: nil)
    }
    
    //MARK: - WkNavigationDelegate
    
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        if navigationAction.request.url?.scheme == "disqus-redirect" {
            decisionHandler(.cancel)
            callback(navigationAction.request.url!.query!.replacingOccurrences(of: "code=", with: ""))
            self.dismiss(animated: true, completion: nil)
        } else {
            decisionHandler(.allow)
        }
    }

}


Sources/DisqusService.swift
//
//  DisqusService.swift
//  Subspedia
//
//  Created by Matteo Riva on 05/09/16.
//  Copyright © 2016 Matteo Riva. All rights reserved.
//

import Foundation
import SafariServices

public extension Notification.Name {
    static let DisqusServiceSafariAuthDidClose = Notification.Name("DisqusServiceSafariAuthDidClose")
}

public class DisqusService: NSObject, SFSafariViewControllerDelegate {
    
    public typealias disqusAuthCompletion = (Bool) -> Void
    public typealias disqusAPICompletion = ([AnyHashable : Any]?,Bool) -> Void
    
    public static let shared = DisqusService()
    
    private let authURL = "https://disqus.com/api/oauth/2.0/"
    private let baseURL = "https://disqus.com/api/3.0/"
    
    private var secretKey: String!
    private var publicKey: String!
    private var redirectURI: String!
    
    private var loggedUser: DisqusUser? {
        didSet {
            if loggedUser != nil {
                let data = NSKeyedArchiver.archivedData(withRootObject: loggedUser!)
                UserDefaults.standard.set(data, forKey: "disqusLoggedUser")
            } else {
                UserDefaults.standard.removeObject(forKey: "disqusLoggedUser")
            }
        }
    }
    
    public var loggedUserID: String? {
        get { return loggedUser?.userID }
    }
    
    public var isUserAuthenticated: Bool {
        get { return loggedUser != nil }
    }
    
    //MARK: - Init
    
    private override init() {
        if let data = UserDefaults.standard.data(forKey: "disqusLoggedUser") {
            loggedUser = NSKeyedUnarchiver.unarchiveObject(with: data) as? DisqusUser
        }
        super.init()
    }
    
    public func set(publicKey: String, secretKey: String, redirectURI: String) {
        self.publicKey = publicKey
        self.secretKey = secretKey
        self.redirectURI = redirectURI
        loggedUser?.refreshToken(publicKey: publicKey, secretKey: secretKey) {[unowned self] (success) in
            if success {
                let data = NSKeyedArchiver.archivedData(withRootObject: self.loggedUser!)
                UserDefaults.standard.set(data, forKey: "disqusLoggedUser")
            }
        }
    }
    
    //MARK: - Auth
    
    public func authenticate(viewController: UIViewController, completionHandler: @escaping disqusAuthCompletion) {
        
        var urlString = "authorize/"
        urlString += "?client_id=\(publicKey!)"
        urlString += "&scope=read,write"
        urlString += "&response_type=code"
        urlString += "&redirect_uri=\(redirectURI!)"
        
        let url = URL(string: authURL + urlString)!
        
        if #available(iOS 9.0, *) {
            let safariVC = SFSafariViewController(url: url)
            safariVC.delegate = self
            viewController.present(safariVC, animated: true, completion: nil)
            NotificationCenter.default.addObserver(forName: .DisqusServiceSafariAuthDidClose,
                                                   object: nil, queue: .main ) {[unowned self] (notif) in
                                                    safariVC.dismiss(animated: true, completion: nil)
                                                    let tmpCode = (notif.object as! URL).query!.replacingOccurrences(of: "code=", with: "")
                                                    let url2 = URL(string: self.authURL + "access_token/")!
                                                    let params = ["grant_type" : "authorization_code",
                                                                  "client_id" : self.publicKey!,
                                                                  "client_secret" : self.secretKey!,
                                                                  "redirect_uri" : self.redirectURI!.addingPercentEncoding(withAllowedCharacters: .alphanumerics)!,
                                                                  "code" : tmpCode]
                                                    
                                                    self.performPOSTConnection(url: url2, parameters: params) { [unowned self] (data, error) in
                                                        if let json = data as? [AnyHashable : Any] {
                                                            self.loggedUser = DisqusUser(json: json)
                                                        }
                                                        completionHandler(error == nil)
                                                    }
            }
        } else {
            let nav = DisqusAuthViewController()
            nav.url = url
            nav.callback = { (tmpCode) in
                let url2 = URL(string: self.authURL + "access_token/")!
                let params = ["grant_type" : "authorization_code",
                              "client_id" : self.publicKey!,
                              "client_secret" : self.secretKey!,
                              "redirect_uri" : self.redirectURI!.addingPercentEncoding(withAllowedCharacters: .alphanumerics)!,
                              "code" : tmpCode]
                
                self.performPOSTConnection(url: url2, parameters: params) { [unowned self] (data, error) in
                    if let json = data as? [AnyHashable : Any] {
                        self.loggedUser = DisqusUser(json: json)
                    }
                    completionHandler(error == nil)
                }
            }
            let navVC = UINavigationController(rootViewController: nav)
            viewController.present(navVC, animated: true, completion: nil)
        }
    }
    
    public func logout() {
        loggedUser = nil
    }
    
    //MARK: - Api call
    
    public func performGETRequest(api: String, authRequired: Bool = false, params: [AnyHashable : Any], completionHandler: @escaping disqusAPICompletion) {
        performRequest(api: api, authRequired: authRequired, params: params, method: "GET", completionHandler: completionHandler)
    }
    
    public func performPOSTRequest(api: String, authRequired: Bool = false, params: [AnyHashable : Any], completionHandler: @escaping disqusAPICompletion) {
        performRequest(api: api, authRequired: authRequired, params: params, method: "POST", completionHandler: completionHandler)
    }
    
    private func performRequest(api: String, authRequired: Bool, params: [AnyHashable : Any], method: String, completionHandler: @escaping disqusAPICompletion) {
        let url = URL(string: baseURL + api + ".json")!
        
        var params = params
        params["api_key"] = publicKey!
        params["api_secret"] = secretKey!
        
        if let token = loggedUser?.accessToken {
            if authRequired {
                params["access_token"] = token
            }
        }
        
        let block: conCompletion = { (data, error) in
            let errorCond = error == nil && ((data as? [AnyHashable : Any])?["code"] as? Int) == 0
            completionHandler(data as? [AnyHashable : Any], errorCond)
        }
        
        if method == "GET" {
            performGETConnection(url: url, parameters: params, completionHandler: block)
        } else if method == "POST" {
            performPOSTConnection(url: url, parameters: params, completionHandler: block)
        }
        
    }
    
    //MARK: - General
    
    private typealias conCompletion = (Any?, Error?) -> Void
    
    private func performGETConnection(url: URL, parameters: [AnyHashable : Any]?, completionHandler: @escaping conCompletion) {
        
        var paramString = "?"
        if let parameters = parameters {
            
            for (key,value) in parameters {
                paramString += "\(key)=\(value)&"
            }
            paramString.characters.removeLast()
        }
        let url = URL(string: url.absoluteString + paramString)!
        
        performConnection(url: url, method: "GET", parameters: nil, completionHandler: completionHandler)
    }
    
    private func performPOSTConnection(url: URL, parameters: [AnyHashable : Any]?, completionHandler: @escaping conCompletion) {
        performConnection(url: url, method: "POST", parameters: parameters, completionHandler: completionHandler)
    }
    
    private func performConnection(url: URL, method: String, parameters: [AnyHashable : Any]?, completionHandler: @escaping conCompletion) {
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if method == "POST" {
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            
            if let parameters = parameters {
                
                var paramString = ""
                for (key,value) in parameters {
                    paramString += "\(key)=\(value)&"
                }
                paramString.characters.removeLast()
                
                request.httpBody = paramString.data(using: String.Encoding.utf8)!
            }
        }
        
        URLSession.shared.dataTask(with: request, completionHandler: { (data, _, error) in
            
            var json: Any? = nil
            if data != nil {
                json = try? JSONSerialization.jsonObject(with: data!, options: [])
            }
            
            completionHandler(json, error)
        }).resume()
    }
    
    //MARK: - SFSafariViewControllerDelegate
    
    @available(iOS 9.0, *)
    public func safariViewControllerDidFinish(_ controller: SFSafariViewController) {
        
    }

}
